### ISSUE TITLE
Tutorial: Instacart

### ISSUE THREAD

**Issue by crockpotveggies** [deeplearning4j/deeplearning4j-docs#36]
## Due Date
*To be completed by:* 2018-05-10

## Description
A tutorial using instacart data and outlining the standard approaches vs. multi task.

https://www.kaggle.com/c/instacart-market-basket-analysis

## Assignees
*Please ensure you have assigned at least one person to this issue. Include any authors and reviewers required.*

**Comment by bpark738** [comment]
@turambar Yeah basically the goal was to read in a sequence of indicators whether a user bought a product in a aisle (134 in total) for a certain order. So the length of the sequence is the total number of orders for a user (minus one, since we are predicting the last order). 

The task is to then predict whether a user buys a product from a certain department in the next order. I tried frozen foods, snacks and deli and made 3 separate networks, which gave AUC's around 0.75.

Then i tried multitasking with a computationgraph using 3 output layers connected to a graveslstm layer. It got like AUC's of around 0.61-0.64. I also tried multitasking with only 2 of the 3 outputs at a time but that also gave AUC's around 0.6's.

**Comment by bpark738** [comment]
maybe multitaskin would work better if the outputs were somehow related?  i chose deli ,snacks, and frozen food since they had a lot of purchases. there's 20 departments total 

1,frozen
2,other
3,bakery
4,produce
5,alcohol
6,international
7,beverages
8,pets
9,dry goods pasta
10,bulk
11,personal care
12,meat seafood
13,pantry
14,breakfast
15,canned goods
16,dairy eggs
17,household
18,babies
19,snacks
20,deli

**Comment by turambar** [comment]
@bpark738 well that is the conventional wisdom about multitask learning (MTL): you want tasks that are related but maybe not TOO related (it's not clear whether MTL with, say, ten identical tasks would do anything).

That said, in the Stanford project, we haven't been able to find any empirical evidence for the conventional wisdom. When we choose ~10 auxiliary tasks from among 1000 candidates, we found no difference when choosing them at random vs. choosing tasks that are highly correlated with the target task.

**Comment by bpark738** [comment]
Hm yeah I'm getting similar things to your stanford project.I tried a bunch of combinations that had the highest correlations and it still did a lot worse than single task. 

What angle should the post take on the multi-learning stuff if at all? Idt the instacart stuff adds anything new if the multi-learning is excluded though.

**Comment by bpark738** [comment]
i changed the problem a bit and got a small improvement using multitask learning. 

So instead of having one target at the end, i used a target sequence (indicator of whether or not a user bought a product in a certain department in the next order relative to the order at the current time step). then evaluated the model at each step.  should have done that to start with.

I used dairy and breakfast departments and for multitasking got around 0.769 AUC for breakfast and for single task for breakfast got around 0.76. Dairy did a bit worse compared to single task.
